src/Buckle/Buckle/CodeAnalysis/Evaluator.cs

```diff
@@ -145,23 +229,33 @@ internal sealed class Evaluator {
         Assign(statement.variable, value);
     }
 
-    private void Assign(VariableSymbol variable, object value) {
+    private void Assign(VariableSymbol variable, EvaluatorObject value) {
         if (variable.type == SymbolType.GlobalVariable) {
-            globals_[variable] = value;
+            var currentValue = globals_.ContainsKey(variable) ? globals_[variable] : null;
+
+            if (currentValue != null && currentValue.isReference && !value.isReference)
+                Assign(currentValue.reference, value);
+            else
+                globals_[variable] = value;
         } else {
             var locals = locals_.Peek();
-            locals[variable] = value;
+            var currentValue = locals.ContainsKey(variable) ? locals[variable] : null;
+
+            if (currentValue != null && currentValue.isReference && !value.isReference)
+                Assign(currentValue.reference, value);
+            else
+                locals[variable] = value;
```

Comment:
If a variable is a reference, it assigns to what it is referencing unless the new value is a new reference.